Implement debouncing of diagnostics

cocreature opened the pull request:
Previously, we emitted diagnostics notifications as soon as we got
them. This resulted in a lot of flickering due to diagnostics getting
cleared briefly when typing only to immediately reappear.

Now, we buffer them for 0.1s so that a new event restoring the
same diagnostics for a slightly modified file will overwrite the
initial clear of diagnostics for the new document version.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

neil-da commented on compiler/hie-core/src/Development/IDE/Core/Debouncer.hs:
```diff
@@ -0,0 +1,67 @@
+-- only the new event will fire.
+newtype Debouncer k = Debouncer (Var (Map k (Async ())))
+
+newtype MicroSeconds = MicroSeconds Int
```
My preference is just to always use System.Time.Extra.Seconds for all seconds, since people don't think in microseconds very naturally and always get it wrong.

cocreature commented:
Changed to use `System.Time.Extra.Seconds` and I changed the delay to be `0` if the list of diagnostics is non-empty. That should give us the behavior you are describing.

neil-da commented on compiler/hie-core/src/Development/IDE/Core/Debouncer.hs:
```diff
@@ -0,0 +1,64 @@
+-- | Wait for all registered events to fire.
+waitForEvents :: Debouncer k -> IO ()
+waitForEvents (Debouncer d) =
+    withVar d $ \m -> mapM_ wait $ Map.elems m
```
This is OK, but a little sad that you always wait the 0.1 even though you know you aren't getting any new things coming in. Not easy to make it better without making it more complex though. Best I can suggest is having a `(Async (), IO ())` where the `IO ()` is a `once` that both try running. However, that's a lot of complexity for something not particularly valuable.

cocreature replied:
I’ve just removed `waitForEvents` since we don’t actually use it.

neil-da replied:
Oh, even better!

neil-da commented on compiler/hie-core/src/Development/IDE/Core/Debouncer.hs:
```diff
@@ -0,0 +1,64 @@
+
+-- | Kill all registered events and clear the map.
+killEvents :: Debouncer k -> IO ()
+killEvents (Debouncer d) = modifyVar_ d $ \m -> do
```
This is never used. Why have it?

cocreature replied:
Removed, I wrote the API for the debouncer before implementing it and thought it might turn out to be useful.

neil-da commented on compiler/hie-core/src/Development/IDE/Core/Shake.hs:
```diff
@@ -391,14 +394,16 @@ updateFileDiagnostics fp k ShakeExtras{diagnostics, state, eventer} current = li
-            eventer $ publishDiagnosticsNotification fp newDiags
+        let uri = fromNormalizedUri $ filePathToUri' fp
+        when (newDiags /= oldDiags) $ do
+            let delay = if null newDiags then 0.1 else 0
```
Thanks!
